storage/src/key_value/agent.rs

```diff
@@ -0,0 +1,603 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{any::Any, collections::VecDeque};
+
+use snarkvm_dpc::{BlockHeaderHash, StorageError};
+use snarkvm_utilities::{has_duplicates, FromBytes, ToBytes};
+use tokio::sync::mpsc;
+
+use crate::{
+    key_value::*,
+    BlockStatus,
+    CanonData,
+    Digest,
+    ForkDescription,
+    ForkPath,
+    KeyValueStorage,
+    SerialBlock,
+    SerialBlockHeader,
+    SerialTransaction,
+    TransactionLocation,
+};
+
+pub(super) struct Agent<S: KeyValueStorage + 'static> {
```

Comment:
This could use a more descriptive name, even if just `StorageAgent`.

Reply:
it's not a storage agent, but an agent for `key_value` -- ditto context-eliminated struct names